bundler/exe/bundle

```diff
@@ -18,14 +18,15 @@ end
 # Workaround for non-activated bundler spec due to missing https://github.com/rubygems/rubygems/commit/4e306d7bcdee924b8d80ca9db6125aa59ee4e5a3
 gem "bundler", Bundler::VERSION if Gem.rubygems_version < Gem::Version.new("2.6.2")
 
-# Check if an older version of bundler is installed
-$LOAD_PATH.each do |path|
-  next unless path =~ %r{/bundler-0\.(\d+)} && $1.to_i < 9
-  err = String.new
-  err << "Looks like you have a version of bundler that's older than 0.9.\n"
-  err << "Please remove your old versions.\n"
-  err << "An easy way to do this is by running `gem cleanup bundler`."
-  abort(err)
+if RUBY_VERSION < "2.6.0.a" && !ENV["BUNDLER_NO_OLD_RUBIES_WARNING"]
+  Bundler.ui.warn \
+    "Running #{Bundler::VERSION} on Ruby #{RUBY_VERSION} is deprecated, " \
```

Comment:
:thinking: This could print wrong info on non CRuby implementations. For example on JRuby 9.2.20.1 (released 1st of December this year) `RUBY_VERSION` is `2.5.7` and AFAIK it is not EOL.

Maybe we should check if CRuby is used and print this detailed info. And fallback to more general message when on other implementation.

@enebo @eregon any ideas?

Reply:
Oh, nice catch! The good news is that `JRuby` is shipping a recent enough rubygems version, so `gem install bundler` won't break there when we drop support for old rubies. So I think it would be fine to only bug CRuby users.

Reply:
Well, thanks actually incorrect, that improvement was released with rubygems 3.2.3, and jruby 9.2.20.1 ships with 3.1.6. So... forget about what I said.

Reply:
I'm thinking of skipping the warning on either Ruby 2.5 or JRuby for now. Not sure if JRuby plans to release one more 9.2 version. If they could release one last version including updated RubyGems, the would be ideal and we could skip this warning on JRuby.

Reply:
cc @headius 
The part about "reached its End of Life and no longer receives security fixes" is only true for CRuby.

I feel "EOL Rubies" here are used as a reason to not support them in Bundler anymore, but I think the real reason is Bundler doesn't want to support older Ruby versions because it occurs extra maintenance efforts (that makes sense).
I feel it's not Bundler's place to warn about EOL Rubies, the last gem which did this (incorrectly) was very annoying: https://github.com/e2/ruby_dep

IMHO, if gems want a minimum version, they should (just) use `required_ruby_version` (and a new major or minor version).
I get a warning is useful here given the issue you mention with RubyGems.
So I think the message should be more about Bundler's choice for a minimum Ruby version than any Ruby EOL or anything like that.

Reply:
i.e., I think the message should be like "Bundler x.y will require Ruby >= a.b; Due to a bug in RubyGems < c.d `gem install bundler` will not work correctly and `gem install bundler:'< 2.3.0'` is needed. Please update your setup to use bundler < 2.3.0 when running on Ruby x.y.z".

Reply:
You are right about the maintenance reason to drop support, but not using insecure versions is also another reason, much easier to explain to users.

Your suggestion feels too complicated to me, specially since users might actually be running a proper rubygems version when they get it. We could skip the warning in that case, but I think we want all users of old rubies to read this, because they might still have scripts, or CI configurations subject to the `gem install bundler` problem.

I feel focusing on old rubies is simpler to explain and understand.

Reply:
> I feel it's not Bundler's place to warn about EOL Rubies, the last gem which did this (incorrectly) was very annoying: https://github.com/e2/ruby_dep

The way I see it, part of the idea behind doing this is to be annoying 😆. Be a bit annoying now, so that we avoid being really annoying later when a lot of things break when we drop support.

Reply:
> Your suggestion feels too complicated to me, specially since users might actually be running a proper rubygems version when they get it.

I think my suggestion is simply the right thing to do, anything else feels inappropriate (it's not Bundler's job to tell people about old Ruby versions, they're very likely already aware).

> We could skip the warning in that case

I think it's important to only warn for this specific case and no other.

> , but I think we want all users of old rubies to read this, because they might still have scripts, or CI configurations subject to the `gem install bundler` problem.

Those people already know they are on old versions, they'll just be annoyed to see a warning they typically can't do anything about (i.e., if they already have a recent enough RubyGems).
If they plan to migrate to a newer Ruby it's unlikely Bundler will be the only/main reason for it.

> I feel focusing on old rubies is simpler to explain and understand.

It's also completely hiding the real issue and reason for this warning.
For me it's a clear case of don't assume users are dumb, let them make their choices, explain the problem clearly, and don't use something else as an excuse.

Bundler wants to drop support for Ruby < 2.6, that's fine, but Bundler should own it and explain why Bundler warns here (and not just change `required_ruby_version`), not use something which is not the real reason.

Reply:
> The way I see it, part of the idea behind doing this is to be annoying laughing. Be a bit annoying now, so that we avoid being really annoying later when a lot of things break when we drop support.

It's good to warn when people need to do something about it and to avoid breakage. It's bad to warn in other cases.
And the reason needs to be clear, "you're using EOL Rubies" will I think only annoy people but they typically won't do anything about it if that's the main thing mentioned in the warning.
"Bundler x.y will drop support for Ruby < 2.6", that's something everyone can understand and makes sense, there is no "you should not use EOL Rubies" (I've tried the EOL argument for setup-ruby, it's always a waste of time, some people made the choice to use an EOL Ruby, the only thing that works is to set required Ruby versions for gems).

Reply:
> I think my suggestion is simply the right thing to do, anything else feels inappropriate (it's not Bundler's job to tell people about old Ruby versions, they're very likely already aware).

I can see how your suggestions very usually feel like the right thing to do to you 🤣. It happens to all of us.

Anyways, you have a point here, I'll try to reword the warning to be easy to understand and yet properly explain things.

Reply:
:D
Another way to phrase this:
What is the most likely to get users to change to `gem install bundler:'< 2.3.0'` when needed?
* To tell them they use an EOL Ruby version and more things
* To tell them it'll break with Bundler x.y if they don't do it due to an issue in RubyGems a.b

Reply:
It should suggest `< 2.3.0` as that then addresses the warning.

> Also, should we show the message for users running a "safe" version of rubygems? Maybe we can skip it there?

Yes I believe we should skip it if there is no need to change anything and it just works, and just verify it indeed works as expected on those RubyGems versions (probably already checked in tests?).

Reply:
If they are already running a higher version of `bundler`, that won't address the warning because they will most likely keep using that version by default. I think the best thing to suggest is actually `gem update --system 3.3.0`. That will address the issue and make the warning not reappear since they will now have a modern enough rubygems version. Suggesting an upgrade is also better for us maintainers than suggesting a downgrade.

Reply:
> IMHO, if gems want a minimum version, they should (just) use required_ruby_version (and a new major or minor version).

That's actually next step once deprecation message would be around long enough. This message is just developer friendly semi-step in between.

- `<=` 2.2.x series is the last one supported on old Rubies without any warning
- `=` 2.3.x is last series supporting old Rubies, but with warning (maybe before 2.4+ release we can release last 2.3 without this warning)
- `>=` 2.4.x will drop support for old Rubies (and if possible will make it clear in `required_ruby_version`)

Reply:
@simi What do you think about focusing on the rubygems version rather than on the ruby version? I think it's better because it's more clear, and also it's independent from whether `RUBY_VERSION` is considered unsupported or not for specific implementations.

How about something like this

> Your rubygems version (x.y.z) has a bug that makes dropping support for old rubies in bundler through bumping the `required_ruby_version` field in the bundler gemspec unsafe for any scripts including `gem install bundler`. That means that, if those scripts are using rubygems x.y.z, they will break as soon as bundler releases a version dropping support for your ruby version (a.b.c). Please upgrade rubygems through `gem update --system` to avoid future breakage.

Reply:
Why downgrading a software, and thus giving up on all improvements and bug fixes, would be better than upgrading a different software, and thus opting in to all improvements and bug fixes? We run a full rubygems CI for every commit against all supported versions, including ruby 2.3, ruby 2.4, and ruby 2.5.

I think you should add support for upgrading rubygems in ruby/setup-ruby, and ideally ship a version higher than 3.2.3 on old rubies, so that users that do `gem install bundler` are safe.

Reply:
> Why downgrading a software, ... would be better than upgrading a different software

Because either way Bundler < 2.3 is needed on those Rubies, so anyway there is a downgrade or pinning involved.
And then the difference is installing a new RubyGems including its new features/fixes/changes/bugs or not.

(what will happen when Bundler 2.3.0 is installed with RubyGems 3.3.0? Does it ignore that gem, or does it use it because it's more recent and so would still warn and would need uninstall as well to not warn?)

If people are on EOL Rubies I'd think they prefer to change as little as possible.
Also people might not like to update RubyGems (e.g. I never do it, and TruffleRuby doesn't support it because it's slow in CI, introducing instability/harder to reproduce, and too risky for too little gains, and some CIs might also not support updating RubyGems because it means writing to the Ruby prefix). I guess you already know how I feel about `gem update --system` installing Bundler in `site_ruby`.

> I think you should add support for upgrading rubygems in ruby/setup-ruby

I disagree, but people can do it themselves and probably in the future it'll possible to still use bundler-caching with that (by having a separate action for bundler caching).

> ideally ship a version higher than 3.2.3 on old rubies, so that users that do gem install bundler are safe.

I can already say this will not happen because it'd be a lot of work and it would differ for people just installing Ruby locally which would be a reproducibility issue.

Reply:
Why would we print a message like

> Versions of software X older than a.b.c have a bug that might affect you, please downgrade software Y to a version older than d.e.f

rather than

> Versions of software X older than a.b.c have a bug that might affect you, please upgrade software X to a version higher than or equal to a.b.c

Rephrasing your own arguments, it feels like another case of assuming your users are dumb and hiding the real motivations for things.

Reply:
I guess you wrote this before my reply. I think my reply before addresses that, either way it's a downgrade of Bundler to < 2.3, hence the first is clearer but also a smaller and safer change.

Reply:
> Because either way Bundler < 2.3 is needed on those Rubies, so anyway there is a downgrade or pinning involved.

That doesn't make any sense, if we print this warning when running bundler 2.3 on those rubies, Bundler < 2.3 is definitely not needed.

> If people are on EOL Rubies I'd think they prefer to change as little as possible.

Yet, the people we're targeting here are installing the latest bundler. I don't think this kind of assumptions are accurate.

> Also people might not like to update RubyGems (e.g. I never do it, and TruffleRuby doesn't support it because it's slow in CI, introducing instability/harder to reproduce, and too risky for too little gains, and some CIs might also not support updating RubyGems because it means writing to the Ruby prefix). I guess you already know how I feel about gem update --system installing Bundler in site_ruby.

Alright! Now I see your real motivation for recommending downgrading bundler rather than upgrading rubygems. It seems that you're not following your own motto of being honest to users and not hiding your actual motivations 😅.

Anyways, I think we're going to start going round in circles at this point. My last proposal seems fine to me, if @simi and other maintainers agree I think it's fine to move on with that.

Reply:
> I disagree, but people can do it themselves and probably in the future it'll possible to still use bundler-caching with that (by having a separate action for bundler caching).

> I can already say this will not happen because it'd be a lot of work and it would differ for people just installing Ruby locally which would be a reproducibility issue.

Alright, I respect your decisions about what to do with `ruby/setup-ruby`. Thanks for maintaining it by the way, it's great.

Reply:
> That doesn't make any sense, if we print this warning when running bundler 2.3 on those rubies, Bundler < 2.3 is definitely not needed.

It is to avoid the warning and also make it clear to use a bundler version supporting that old Ruby version.
One subtlety is by updating RubyGems you'd be able to use bundler >= 2.3 and < 2.4, I didn't consider that.
Not sure how much that really matters.
Either way Bundler >= 2.4 can't be used on those older Rubies, hence Bundler will be effectively pinned one way or another.

> Alright! Now I see your real motivation for recommending downgrading bundler rather than upgrading rubygems. It seems that you're not following your own motto of being honest to users and not hiding your actual motivations 

I don't think I was hiding anything, I have many reasons to never update RubyGems, I just tried to present the ones relevant to this discussion. Not being able to reproduce when e.g. `ruby-build/install x.y.z` is maybe the biggest of them in this context.

I also remember when people were updating RubyGems for Bundler 2 even though it wasn't necessary, and I wrote https://eregon.me/blog/2020/01/13/a-migration-path-to-bundler2.html#the-rubygems-requirement-of-bundler-2 about it. I remember various issues and confusing docs on TravisCI related to that.
I think it would be wiser to not only suggest people to update RubyGems, but tell them to either update RubyGems or pin Bundler to < 2.3.0 (for old Rubies).
(For me the second is better but I can see others have different opinions on the matter)

Reply:
Well, as per https://github.com/rubygems/rubygems/pull/5177#discussion_r772481039 you're suggesting a solution different from the straightforward one (upgrade rubygems to fix a bug in rubygems), and you're not explaining the reasons behind it (upgrading rubygems is "bad").

I'm leaning towards only suggesting a rubygems upgrade. I think any maintainer would sympathize with preferring to suggest upgrading the software that you maintain that to suggest downgrading the software that you also maintain. The bundler downgrade is also more complicated, not only conceptually, but because it doesn't remove the warning by itself, you also need to uninstall the currently running bundler, and that might not be straightforward for some users, so it would probably need further clarification.